perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/VosManagerBlImpl.java

```diff
@@ -902,14 +929,86 @@ public boolean usesEmbeddedGroupRegistrations(PerunSession sess, Vo vo) {
+					if (member.getStatus() == Status.EXPIRED) {
+						// member is expired in memberVo but exists in parentVo, don't update
+						continue;
+					}
+
+					// member is valid in memberVo, so reset expiration in parentVo
+					perunBl.getAttributesManagerBl().removeAttribute(sess, existingMember, memberExpirationAttrDef);
+					if (existingMember.getStatus() != Status.VALID) {
+						perunBl.getMembersManagerBl().validateMemberAsync(sess, member);
+					}
+					//update memberOrganizations
+					Attribute attribute = perunBl.getAttributesManagerBl().getAttribute(sess, existingMember, A_MEMBER_DEF_MEMBER_ORGANIZATIONS);
+					ArrayList<String> currentValue = attribute.valueAsList();
+					currentValue.add(memberVo.getShortName());
+					attribute.setValue(currentValue);
+					perunBl.getAttributesManagerBl().setAttribute(sess, existingMember, attribute);
+
+				} catch (MemberNotExistsException e) {
+					// if user is member only in member vo, create member in parent vo
+					Member newMember = perunBl.getMembersManagerBl().createMember(sess, vo, perunBl.getUsersManagerBl().getUserByMember(sess, member));
+					if (member.getStatus() == Status.VALID) {
+						// remove expiration set according to parentVo and update memberOrganizations
+						perunBl.getAttributesManagerBl().removeAttribute(sess, newMember, memberExpirationAttrDef);
+						Attribute attribute = new Attribute(perunBl.getAttributesManagerBl().getAttributeDefinition(sess, A_MEMBER_DEF_MEMBER_ORGANIZATIONS));
+						ArrayList<String> newValue = new ArrayList<>(List.of(memberVo.getShortName()));
+						attribute.setValue(newValue);
+						perunBl.getAttributesManagerBl().setAttribute(sess, newMember, attribute);
+					}
+					perunBl.getMembersManagerBl().validateMemberAsync(sess, member);
+				}
+
+			} catch (WrongAttributeAssignmentException | WrongReferenceAttributeValueException | AttributeNotExistsException | WrongAttributeValueException | AlreadyMemberException | ExtendMembershipException e) {
+				throw new InternalErrorException(e);
+			}
+		}
 	}
 
 	@Override
 	public void removeMemberVo(PerunSession sess, Vo vo, Vo memberVo) throws RelationNotExistsException {
-		// todo - add necessary logic for removing member vo
 		vosManagerImpl.removeMemberVo(sess, vo, memberVo);
+
+		List<Member> parentVoMembers = perunBl.getMembersManagerBl().getMembers(sess, vo);
+
+		for (Member member : parentVoMembers) {
+			try {
+				Attribute attribute = perunBl.getAttributesManagerBl().getAttribute(sess, member, A_MEMBER_DEF_MEMBER_ORGANIZATIONS);
+				ArrayList<String> currentValue = attribute.valueAsList();
+				if (currentValue != null && currentValue.contains(memberVo.getShortName())) {
+					currentValue.remove(memberVo.getShortName());
+					attribute.setValue(currentValue);
+					perunBl.getAttributesManagerBl().setAttribute(sess, member, attribute);
+				}
+				if (currentValue != null && currentValue.equals(new ArrayList<>(List.of(memberVo.getShortName())))) {
```

Comment:
```suggestion
				if (currentValue != null && currentValue.equals(new ArrayList<>(List.of(vo.getShortName())))) {
```
I think you meant to do this.

Reply:
What I wanted to achieve here was to check, if before removing the memberVo from memberOrganizations it was the only one in the list. If so, parentVo would overtake to set expiration. If there was another vo in the list, then it should stay intact i think.

Reply:
I get it now, it indeed should check only parent vo stayed in the memberOrganizations. I added empty list as a possibility too.